[batch, ci, scorecard] build on web_common for unified styling

cseed opened the pull request:
Use web_common in ci, batch and scorecard for common design (e.g. header) and styling.  The styling is clearly not ideal, this is intended to be a work in progress.

Hand-deployed to checked batch and scorecard UI are still workable.  Not easy to do with CI.

Converted scorecard to aiohttp and gidgethub.  Will remove dependencies in separate PR so I don't want to wait for all the images to rebuild.  Can't quite get rid of Flask yet as it is used in callback tests.

akotlar commented on web_common/web_common/styles/main.scss:
```diff
@@ -1,68 +1,101 @@
     display: flex;
-    padding: 0 $margin;
-    align-content: center;
+    flex-direction: row;
```
not needed, this is the default.

akotlar commented on notebook2/notebook/styles/pages/notebook.scss:
```diff
@@ -1,35 +1,76 @@
-      margin-bottom: $margin;
-      justify-content: space-around;
-      min-width: 320px;
+        flex-direction: row;
```
can get rid of this.

akotlar commented on web_common/web_common/styles/main.scss:
```diff
@@ -1,68 +1,101 @@
-  margin: 0px;
-}
-
-#theme-site {
```
Not sure why we got rid of this

cseed replied:
I admit I didn't understand the point of this.  Didn't seem necessary to me.

akotlar replied:
The #theme-site was a top-level identifier, to add more specificity for rules (in case of conflict with styling libraries we may in the future use), and to allow the theme to change (click a button, change theme-site to theme-site-dark or whatever). But I was asking about the rules. defined within theme-site, which defined some base styles

cseed replied:
Ah, I see.  I'm inclined to put it back when I actually need it for an application like you mention.  Changing the id on a container to change the styling is a slick trick.  I'm continually impressed with the stuff you can do that browsers can roll with.

It's all there, the relevant stuff anyway.  I added a `body { ... }` run that has the margin and font stuff.  The layout stuff didn't need to go anywhere, I'm using the default font-size (to be revisited), and white background is also the default.

akotlar commented on web_common/web_common/styles/typography.scss:
```diff
@@ -1,32 +0,0 @@
-@import "variables/colors.scss";
```
again, not sure why this left

cseed replied:
I'm confused, it's deleted.

akotlar replied:
I just mean, why did typeography get deleted? Now the Hail logo is just body font-size.

akotlar commented:
Left some comments. Will finish tomorrow; stats today

akotlar commented on batch/batch/batch.py:
```diff
@@ -1160,8 +1164,9 @@ def is_successful(self):
+    context = base_context(deploy_config, userdata, 'batch')
+    context['batch_list'] = batches[::-1]
+    context['token'] = token
     response = aiohttp_jinja2.render_template('batches.html',
```
For consistency prefer @aiohttp_jinja2.template('batches.html')

cseed replied:
We can't here because we're setting the _csrf token.  I think this is the intended usage, see the "Using a function based web handler" example in the docs:

https://github.com/aio-libs/aiohttp-jinja2

akotlar replied:
Ah interesting.

akotlar commented on build.yaml:
```diff
@@ -160,10 +163,20 @@ steps:
    name: scorecard_image
    dockerFile: scorecard/Dockerfile
-   contextPath: scorecard
+   contextPath: .
```
What is contextPath?

cseed replied:
It's path where the docker build command is run.

akotlar replied:
So this depends on whether the module depends on anything held outside of its root (for instance, something that depends on /auth) since docker cannot copy anything outside of its context. Correct?

cseed replied:
Correct.  I'm not sure how careful we are about always having the lowest possible context path.  `scorecard` might work here again: before I was installing gear in this image, but now I'm putting it in the services base image.

akotlar replied:
>  You mean the "Hail" in the center the / page?

Yes

cseed replied:
I left it because I'm going to turn the / page into a dashboard that has a summary box for each service: link to a running notebook, activate and recent batch jobs, PR status for developers, etc.

akotlar replied:
Right but why get rid of all <hN> styling? You have a bunch of <hN> elements.

akotlar replied:
Right, but why remove all styling on <hN> elements?

> Oops, sorry, I misread where the failure was happening. You need `cseed/hail:batch-web`, not `cseed:batch-web`. I created an issue to fix this: #7074

Not entirely working, but I'm happy to just review what you have in your namespace.

For reference:
```sh
received unexpected HTTP status: 503 Service Unavailable
```

https://ci.hail.is/batches/13547

cseed replied:
> Right, but why remove all styling on elements?

It was just styling for headers. Mainly, I didn't like how the styling you had looked with how I was using headers in the HTML.

I'm slightly confused why you're using a header tag on what you're calling a logo.  I thought headers were mainly used to tag/name sections of a page.

akotlar replied:
Sorry, "heading"

cseed replied:
> received unexpected HTTP status: 503 Service Unavailable

It looks like this is from GCR while doing a docker push.  Surprising, I've never seen it  before.  Is it reproducible?  I bet GCR just had a temporary hiccup.

akotlar commented on scorecard/scorecard/scorecard.py:
```diff
@@ -51,22 +31,40 @@
+@aiohttp_jinja2.template('index.html')
+@web_maybe_authenticated_user
+async def index(request, userdata):  # pylint: disable=unused-argument
+    context = base_context(deploy_config, userdata, 'notebook2')
```
'scorecard'

akotlar commented on scorecard/scorecard/scorecard.py:
```diff
@@ -51,22 +31,40 @@
+@aiohttp_jinja2.template('user.html')
+@web_maybe_authenticated_user
+async def html_get_user(request, userdata):
+    context = base_context(deploy_config, userdata, 'notebook2')
```
'scorecard'

akotlar commented on scorecard/scorecard/scorecard.py:
```diff
@@ -226,57 +233,62 @@ def update_data():
-        log.error('Exceeded rate limit: ' + str(e))
-
+    except Exception:  # pylint: disable=broad-except
+        log.exception('update failed due to except')
```
It would be nice to get the actual exception here.

akotlar replied:
Ah nvm, log.exception I think gives the stack trace, so should show it.

cseed replied:
`log.exception` logs the exception along with the traceback.

akotlar commented:
Sadness, the css is transforming.

akotlar commented:
>  gidgethub. Will remove dependencies in separate PR so I don't want to wait for all the images to rebuild. Can't quite get rid of Flask yet as it is used in callback tests.

Could you help me understand what the benefit of gidgethub is over our previous strategy? Since we generally prefer minimal reps

cseed commented:
> Could you help me understand what the benefit of gidgethub is over our previous strategy? Since we generally prefer minimal deps

This isn't more deps.  I switched gidgethub for PyGithub.

gidgethub is async, PyGithub is not.  Now that we're using aiohttp over Flask, if we don't use an async library to query Github handling web requests will hang while the synchronous library blocks waiting for requests.

Actually, this is fewer deps.  We already use gidgethub in CI, so overall this removes the PyGithub dependency.  Since both CI and scorecard depend on gidgethub, I moved it from the CI-specific installation to the service base image.

cseed commented:
I fixed scorecard deploy stuff and now it is working with `dev deploy`.  I also pushed some CSS changes:
 - body { margin: 0; } that removes the extra header spacing
 - but added an 0 8px 8px 8px margin to #content
 - simplified the header layout CSS
 - fixed the header item clickable area, should be bigger and uniform across header items
 - fixed the misalignment on Safari

I'm pretty happy with this for this iteration.

akotlar commented:
> > Could you help me understand what the benefit of gidgethub is over our previous strategy? Since we generally prefer minimal deps
> 
> This isn't more deps. I switched gidgethub for PyGithub.
> 
> gidgethub is async, PyGithub is not. Now that we're using aiohttp over Flask, if we don't use an async library to query Github handling web requests will hang while the synchronous library blocks waiting for requests.

Thanks, I figured you would do the caching run in a different thread (run_in_executor). Seems like a good solution

akotlar commented:
In trying to test this:

```sh
(hail) alex:~/projects/hail/hail/python:$ hailctl dev deploy -b cseed:batch-web -s deploy_auth,deploy_router,deploy_notebook2
Traceback (most recent call last):
  File "/miniconda3/envs/hail/bin/hailctl", line 11, in <module>
    sys.exit(main())
  File "/Users/alex/projects/hail/hail/python/hailtop/hailctl/__main__.py", line 100, in main
    cli.main(args)
  File "/Users/alex/projects/hail/hail/python/hailtop/hailctl/dev/cli.py", line 52, in main
    cli.main(args)
  File "/Users/alex/projects/hail/hail/python/hailtop/hailctl/dev/deploy/cli.py", line 66, in main
    loop.run_until_complete(submit(args))
  File "/miniconda3/envs/hail/lib/python3.6/asyncio/base_events.py", line 484, in run_until_complete
    return future.result()
  File "/Users/alex/projects/hail/hail/python/hailtop/hailctl/dev/deploy/cli.py", line 57, in submit
    batch_id = await ci_client.dev_deploy_branch(args.branch, steps)
  File "/Users/alex/projects/hail/hail/python/hailtop/hailctl/dev/deploy/cli.py", line 46, in dev_deploy_branch
    self._deploy_config.url('ci', '/api/v1alpha/dev_deploy_branch'), json=data) as resp:
  File "/miniconda3/envs/hail/lib/python3.6/site-packages/aiohttp/client.py", line 1005, in __aenter__
    self._resp = await self._coro
  File "/miniconda3/envs/hail/lib/python3.6/site-packages/aiohttp/client.py", line 581, in _request
    resp.raise_for_status()
  File "/miniconda3/envs/hail/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 942, in raise_for_status
    headers=self.headers)
aiohttp.client_exceptions.ClientResponseError: 500, message='Internal Server Error'
```

akotlar commented:
Checked https://internal.hail.is/cseed/notebook2/notebook. Crash on trying to create a notebook. Minor styling fixes as well (Hail logo is tiny), can be delayed for future PR.

cseed commented:
Did you copy the the oath2 key to your namespace?   You need to do that for the auth service to boot.  This is probably a cascaded failure because auth is down.

cseed commented:
> Minor styling fixes as well (Hail logo is tiny)

Where is the Hail logo tiny?  You mean the "Hail" in the center the / page?

akotlar commented:
> Did you copy the the oath2 key to your namespace? You need to do that for the auth service to boot. This is probably a cascaded failure because auth is down.

I think I have what I need. I have a ~/.hail/token and ~/.hail/tokens.json. The token file looks like a jwt, decodes to "email": "ako" followed by some low-ascii characters. 

Before I ran this I logged in using hailctl auth login.

cseed commented:
> In trying to test this (from your branch, ran pip install on /hail/python just in case)

You're running this locally, or with `hailctl dev deploy`?  I assume the latter because the former is essentially impossible.

~/.hail/token is no longer used and you can delete it.  You'll need a valid tokens.json to run the dev deploy.  Once the dev deploy runs, your local configuration is irrelevant.  It sounds like your dev deploy was successful.

You're getting failures in your deployed services.  You need to look into your namespace to debug them.  In particular, for auth to run, you're going to need a copy of auth-oauth2-client-secret from the production namespace.  To log in inside the dev namespace, you'll have to add the callback to the list of registered callbacks at Google.  You can copy mine as an example.

cseed commented:
> Crash on trying to create a notebook.

I looked into this.  The notebook launch failed because your account in my namespace isn't complete, it can only login, but doesn't have the other secrets necessary to launch a notebook (e.g. gsa-key).

cseed commented:
Oops, sorry, I misread where the failure was happening.  You need `cseed/hail:batch-web`, not `cseed:batch-web`.   I created an issue to fix this: https://github.com/hail-is/hail/issues/7074

cseed commented:
OK, I think I've addressed all your comments.  Unless I missed something, I'm not aware of other changes you're requesting, so I'm dismissing your review.  Let me know if there is anything else.

Jackie was also seeing issues with GCR today also, that 503 is likely a transient issue.
